BESTPRACTICES.md

```diff
@@ -4,17 +4,268 @@
 First of all to understand weave design philosophy and for the sake of programming well designed software, I advise you to read this fine-grained article:
 [Things I Learnt The Hard Way (in 30 Years of Software Development)](https://blog.juliobiason.net/thoughts/things-i-learnt-the-hard-way/) \
 \
-This documents content is curated by the PR discussions of weave tutorial. If you follow the *PRs* section you can see how implement blockchain app using weave.
+This documents content is curated from PR discussions of weave tutorial. If you follow the *PRs* section you can see how to implement blockchain app using weave.
 And you can see how the development flow works and what are the issues you should have on your mind while moving forward.
 
+### Development flow
+1. #### Codec
+    > [PR#1](https://github.com/iov-one/tutorial/pull/1): _Create order book models_
+
+    Codec is the first the component that needs to be designed. Keep in mind that this part is the most important since whole module will depend on *codec*. You can think codec it as *model* in mvc pattern. Yet it is not simple as model. Codec defines the whole application state models and more.
+
+    - #### State
+        ```protobuf
+        message Trade {
+            weave.Metadata metadata = 1;
+            bytes id = 2 [(gogoproto.customname) = "ID"];
+            bytes order_book_id = 3 [(gogoproto.customname) = "OrderBookID"];
+            bytes order_id = 4 [(gogoproto.customname) = "OrderID"];
+            bytes taker = 5 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+            bytes maker = 6 [(gogoproto.casttype) = "github.com/iov-one/weave.Address"];
+            coin.Coin maker_paid = 7;
+            coin.Coin taker_paid = 8;
+            int64 executed_at = 9 [(gogoproto.casttype) = "github.com/iov-one/weave.UnixTime"];
+        }
+        ```
+
+        As you can see above, weave is heavily using `bytes` for identites, addresses etc. At first glance this might seem difficult to handle but if you take a look at [x/orderbook/bucket](https://github.com/iov-one/tutorial/blob/master/x/orderbook/bucket.go#L125) you can see it is very useful for indexing and performance
```

Comment:
Explain magic "ID" field needed if you want to use auto-increment IDs

Reply:
And a short mention of Metadata:
Include this as first field of all objects that can be serialized (in db or on the wire). Needed to allow future migrations (explained later)